### ISSUE TITLE
How to set global ComparingByValue on classes with recursive properties

### ISSUE THREAD

**Issue by garethbudden** [dennisdoomen/fluentassertions#344]
Given the following simple money class (note the Zero property)

```c#
public class Money
{
    private readonly decimal _value;
    private readonly string _currency;

    public Money(decimal value, string currency)
    {
        _value = value;
        _currency = currency;
    }

    // this recursive property causes the issue
    public Money Zero { get { return new Money(0, _currency);} }

    public override bool Equals(object obj)
    {
        var money = obj as Money;

        return money != null
                   && money._currency == _currency
                   && money._value == _value;
    }

    public override int GetHashCode()
    {
        return GetType().GetHashCode() * _currency.GetHashCode() * _value.GetHashCode();
    }
}
```

This test
```c#
[Test]
public void FailsWithMaximumRecursionError()
{
    // fails
    AssertionOptions.AssertEquivalencyUsing(x => x.ComparingByValue<Money>());

    new Money(12, "GBP").ShouldBeEquivalentTo(new Money(12, "GBP"));
}
```
fails with...
> The maximum recursion depth was reached.  The maximum recursion depth limitation prevents stack overflow from occuring when certain types of cycles exist in the object graph or the object graph's depth is very high or infinite.  This limitation may be disabled using the config parameter.

> The member access chain when max depth was hit was: Zero.Zero.Zero.Zero.Zero.Zero.Zero.Zero.Zero.Zero.Zero

but the next test passes - we're just not setting the ComparingByValue globally and instead on the test

```c#
[Test]
public void Passes()
{
    new Money(12, "GBP").ShouldBeEquivalentTo(new Money(12, "GBP"), x => x.ComparingByValue<Money>());
}
```

How should I configure fluent assertions so that the Money class can be configured globally?

**Comment by dennisdoomen** [comment]
Sure, just change the global settings using `AssertionOptions.AssertEquivalencyUsing(x => x.ComparingByValue<Money>())`

**Comment by garethbudden** [comment]
I tried that in the second  code snippet,  but it fails with "maximum recursion depth was reached".  Unless I'm still doing something wrong?

**Comment by dennisdoomen** [comment]
Hmm, that should work. In any case, you need to tell FA that `Money` should be treated as a value type. You can do that globally [like this](https://github.com/dennisdoomen/fluentassertions/blob/e240c8c8d02160fd69e22d7590901df429f86bd3/Tests/FluentAssertions.Shared.Specs/AssertionOptionSpecs.cs#L133).

**Comment by dennisdoomen** [comment]
Fixed by #350.
